### PR TITLE
Roughness

### PR DESCRIPTION
**Pull request by yqin43**
## Description
Added code for adding tags and code for calculating average roughness value of the geometry

## Changes
Updated get_area_triangle function so it can calculate area of a single triangle
Tagged local roughness
Added generic code for adding tags
Added code and test for average roughness value

### CODE REVIEW

File: dagmc_stats/dagmc_stats.py
Comment (kkiesling) [inline dagmc_stats/dagmc_stats.py:278]: A test needs to be added to reflect this change. Currently only no `tris` argument is tested, but one needs to be added that tests if a subset of `tris` is added.

Comment (kkiesling) [reply dagmc_stats/dagmc_stats.py:278]: Also on that test, we should be checking the values of `obs` vs `exp`, not just the length of the data returned.

File: dagmc_stats/dagmc_stats.py
Comment (kkiesling) [inline dagmc_stats/dagmc_stats.py:474]: specify what the keys and values should be in this dictionary.

File: dagmc_stats/dagmc_stats.py
Comment (kkiesling) [inline dagmc_stats/dagmc_stats.py:463]: docstring needs the `tag` input added below.

File: dagmc_stats/dagmc_stats.py
Comment (kkiesling) [inline dagmc_stats/dagmc_stats.py:482]: Right now this will add the tag to the moab instance in memory but it doesn't look like it gets written to a file. I think that in `generate_stats.py` a new file should be written out if tags are added after all the stats calculations are finished. So there needs to be some way to determine if tags have been set. 

One thought is that the value `tag` can be returned from this function in `generate_stats.py`. If after all stats have been run, if `tag` is true, then write the file. Otherwise we don't need to write out a new file. This will allow other tags to be added to the geometry as well and just write out a file with all tags.

File: dagmc_stats/dagmc_stats.py
Comment (kkiesling) [inline dagmc_stats/dagmc_stats.py:495]: this function needs a test- we can discuss if you don't have an idea for how to test this.

File: dagmc_stats/dagmc_stats.py
Comment (kkiesling) [inline dagmc_stats/dagmc_stats.py:510]:
```suggestion
    tag_eh = \
        my_core.tag_get_handle(tag_name, size=1,
                                tag_type=tag_type,
                                storage_type=types.MB_TAG_SPARSE,
                                create_if_missing=True)
```
PEP8

File: dagmc_stats/dagmc_stats.py
Comment (kkiesling) [reply dagmc_stats/dagmc_stats.py:463]: I also think the default should be False

File: dagmc_stats/dagmc_stats.py
Comment (kkiesling) [inline dagmc_stats/dagmc_stats.py:482]:
```suggestion
    
    if tag:
```
recommended blank line for better readability

File: dagmc_stats/dagmc_stats.py
Comment (kkiesling) [reply dagmc_stats/dagmc_stats.py:495]: Also kudos on this function- This will be a good one for reusing in many contexts.

**Review by kkiesling — CHANGES_REQUESTED** [review]
great work! Overall the additions are great. I don't have any major suggestions on what you've added, but I think there are a few things missing still such as some tests and a way to write out a newly tagged geometry.

**Comment by kkiesling** [comment]
Also- python3 test is failing so we need to look into that.

File: tests/test_dagmc_stats.py
Comment (kkiesling) [inline tests/test_dagmc_stats.py:11]: the `test_get_lri()` function should use this new test geometry. I don't think it makes sense for us to use the cube when all the values are zero (I think that is also why the test is failing in python 3).

File: tests/test_dagmc_stats.py
Comment (kkiesling) [inline tests/test_dagmc_stats.py:236]: let's just do the pyramid not the cube for any roughness tests. (see above comment).

File: tests/test_dagmc_stats.py
Comment (kkiesling) [reply tests/test_dagmc_stats.py:11]: also this file can just be called `pyramid.h5m` since there is not a pyramid-1 anywhere.

File: tests/test_dagmc_stats.py
Comment (kkiesling) [inline tests/test_dagmc_stats.py:248]:
```suggestion
        lr_bottom = []
```
pep8

File: tests/test_dagmc_stats.py
Comment (kkiesling) [inline tests/test_dagmc_stats.py:261]: there are a lot of magic numbers in this section. I am mostly wondering what the 25 number is here?

Comment (kkiesling) [reply tests/test_dagmc_stats.py:261]: some comments for what each of these calculations is doing will be helpful to future developers in the event that this test fails.

**Review by kkiesling — CHANGES_REQUESTED** [review]
second review- some notes on your roughness test.

File: tests/test_dagmc_stats.py
Comment (yqin43) [reply tests/test_dagmc_stats.py:261]: 25 is the area of the bottom surface which is a 5*5 square

File: dagmc_stats/dagmc_stats.py
Comment (gonuke) [inline dagmc_stats/dagmc_stats.py:497]: Separation of concerns: Don't call `add_tag` from within this method.  Allow the user to calculate the roughness of vertices and roughness of triangles without changing the mesh DB (adding a tag changes the mesh DB).  By saving this data, the user can then call `add_tag` separately in the method that calls this.

File: dagmc_stats/dagmc_stats.py
Comment (gonuke) [inline dagmc_stats/dagmc_stats.py:490]: Create a separate method that will evaluate roughness for triangles, assuming that the vertex roughness data already exists.

File: dagmc_stats/dagmc_stats.py
Comment (gonuke) [inline dagmc_stats/dagmc_stats.py:502]:
```suggestion
    tri_roughness : (dictionary) the average roughness values for all triangles
```

File: dagmc_stats/dagmc_stats.py
Comment (gonuke) [inline dagmc_stats/dagmc_stats.py:489]: Might be useful if this allowed the user to pass in a list of triangles, like `get_area_triangle()` above.

File: dagmc_stats/dagmc_stats.py
Comment (gonuke) [inline dagmc_stats/dagmc_stats.py:536]: Allow the user to pass in a list of vertices, and use all the vertices only if that list is empty.

File: dagmc_stats/dagmc_stats.py
Comment (gonuke) [inline dagmc_stats/dagmc_stats.py:559]:
```suggestion
        s_i = get_area_triangle(my_core, root_set, geom_dim, tris=adj_tris)
        num += roughness[vert] * sum(s_i) / 3
```

**Review by gonuke — CHANGES_REQUESTED** [review]
A few more suggestions

File: dagmc_stats/dagmc_stats.py
Comment (gonuke) [reply dagmc_stats/dagmc_stats.py:489]: Actually - let's wait on this until we move to Pandas, because at some point we probably want to just populate this for the whole set of vertices

File: dagmc_stats/dagmc_stats.py
Comment (yqin43) [reply dagmc_stats/dagmc_stats.py:536]: Thank you for the suggestion. I made the updates in the get_roughness function instead, so the roughness value can be that of all the vertices in the geometry or the list of vertices that user passes in.

**Review by gonuke — APPROVED** [review]

**Review by kkiesling — APPROVED** [review]
approving so this can be merged

**Comment by gonuke** [comment]
Decided to merge this as is.  The methods are functioning as intended, and the interfaces will be changing anyway as we introduce the class structure.

**Pull request merged**
